### ISSUE TITLE
What is the good method to configure redirection from `/` page to another page ?

### ISSUE THREAD

**Issue by harobed** [gatsbyjs/gatsby#18665]
How can I redirect `/` to `/page-2/` ?

I have tryed 2 solutions:

# First solution

https://github.com/harobed/gatsby-redirect-on-root-index in `master` branch.

When I execute `npm run develop`, I have this error in Browser console:

```
The route "/" matches both a page and a redirect; this is probably not intentional.
```

but no error after `npm build; npm serve`

# Second solution

https://github.com/harobed/gatsby-redirect-on-root-index in `page-resources-is-undefined` branch.

When I execute `npm run develop`, I have this error in Browser console:

```
TypeError: locationAndPageResources.pageResources is undefined
```

When I execute `npm build; npm serve`, I have this error:

```
NOT FOUND
```

this is not found page, redirection don't work.

# Question

What is the good method to configure redirection on `/` page?

**Comment by Pratikkakkad** [comment]
You can simply redirect like this
```
    if (typeof window !== 'undefined') {
      window.location = '/'
    }
```

**Comment by harobed** [comment]
> You can simply redirect like this
> 
> import { navigate } from 'gatsby'
> navigate('/');

@Pratikkakkad  I would like a redirection from `/` to `/page-2` and not the other way round.

Why can't I use this below?

```
    createRedirect({
        fromPath: '/',
        isPermanent: false,
        redirectInBrowser: true,
        toPath: '/page-2/'
    });
```

Why treat / page differently to other pages?

**Comment by harobed** [comment]
@KyleAMathews an idea? what is your vision / guideline about this? It is a bug or a feature?

**Comment by sever1an** [comment]
The message `The route "[...]" matches both a page and a redirect; this is probably not intentional.` shows even if the page is `404.js`, which seems to make this warning pointless as configured.

Redirects seem to generate `page-data.json` requests for the `fromPath` even in production builds.

This might be another issue, but do redirects belong in `match-paths.json` now?

**Comment by sever1an** [comment]
Stopping `gatsby develop` and running `gatsby clean` after deleting `index.js` resolves the `Cannot read property 'page' of undefined` error and restores expected functionality of displaying the dev 404 page.

This is also a bug... deleting a page file should not require restarting `gatsby develop` and really should not require running `gatsby clean`.

**Comment by n-mirzayev** [comment]
I have the same issue. I would expect to pass some option to `createRedirect` that I want only exact match such in react-router, something like this:
 `createRedirect({
        fromPath: '/',
        exactPath: true,
        isPermanent: false,
        redirectInBrowser: true,
        toPath: '/page-2/'
    });`

**Comment by blainekasten** [comment]
👋 @harobed, @n-mirzayev. I've been looking at this a little bit this morning and am not sure I'm seeing what's missing here. 

From my testing, doing a redirect from `/` to `/page-2/` works. You do get a warning, but I'm not sure that's wrong if you have a `src/pages/index`.

What is the expected behavior here?

**Comment by blainekasten** [comment]
@n-mirzayev So to make sure I understand, as it is today, this _works_. But the console message is a bit misleading and incorrect? Is that the problem? Or does something not work.

**Comment by n-mirzayev** [comment]
@blainekasten Functionality works, it redirects as expected. But we get a warning message in console. It can be a question if there is a way to have redirect without getting warning message, otherwise it can be considered as bug and we shouldn't get warning message.

**Comment by sever1an** [comment]
Just checking if the other page is not 404 before showing the error would fix it. Very easy PR

**Comment by blainekasten** [comment]
Great! Thanks for the clarification everyone! 

@sever1an (or @n-mirzayev or anyone else 😄) would you be interested in contributing a fix for this? We love empowering the community to be part of the Gatsby contributor base! Any way I can help get you involved let me know.

**Comment by n-mirzayev** [comment]
@blainekasten I would love to contribute.

**Comment by blainekasten** [comment]
@n-mirzayev great! [This article](https://www.gatsbyjs.org/contributing/setting-up-your-local-dev-environment/) explains how to setup gatsby locally to make changes. It sounds like the expected behavior is to just not do the log if there isn't an index file, or index route.

This log is executed in `gatsby/cache-dir/navigation`

**Comment by mnurik** [comment]
@blainekasten @sever1an Created PR for this issue: https://github.com/gatsbyjs/gatsby/pull/19789 . As this is my first try all comments are welcome :)

**Comment by balibebas** [comment]
:tada: This was resolved in #19789. With the backing PR merged this works without error now:

1. Remove index page and run `gatsby clean`.
1. Add redirect using [`createRedirect`](https://www.gatsbyjs.org/docs/actions/#createRedirect) (Note you'll probably want to set `redirectInBrowser: true` to mimick behavior on client and server.)
1. Restart `gatsby develop` and point browser to site base url.
1. Observe redirect.

**Comment by balibebas** [comment]
:tada: This was resolved in #19789. With the backing PR merged this works without error now:

1. Remove index page and run `gatsby clean`.
1. Add redirect using [`createRedirect`](https://www.gatsbyjs.org/docs/actions/#createRedirect) (you'll probably want to set `redirectInBrowser: true`.)
1. Restart `gatsby develop` and point browser to site base url.
1. Observe redirect.

**Comment by balibebas** [comment]
:tada: This was resolved in #19789. With the backing PR merged this works without error now:

1. Remove index page and run `gatsby clean`.
1. Add redirect using [`createRedirect`](https://www.gatsbyjs.org/docs/actions/#createRedirect) (you'll probably want to set `redirectInBrowser: true`)
1. Restart `gatsby develop` and point browser to site base url.
1. Observe redirect.

**Comment by manikandanvaidy** [comment]
> I have the same issue. I would expect to pass some option to `createRedirect` that I want only exact match such in react-router, something like this:
> 
> ```
>  createRedirect({
>         fromPath: '/',
>         exactPath: true,
>         isPermanent: false,
>         redirectInBrowser: true,
>         toPath: '/page-2/'
>     });
> ```

This got working for me. Thanks @n-mirzayev
